rpcserver.go

```diff
@@ -936,12 +941,128 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
 	)
 	valid := err == nil
 
-	// TODO(roasbeef): also show additional final resting anchor
-	// information, etc?
+	decodedProof, err := marshalProofFile(proofFile, -1)
+	if err != nil {
+		return nil, fmt.Errorf("unable to marshal proof: %w", err)
+	}
+
+	return &tarorpc.VerifyProofResponse{
+		Valid:        valid,
+		DecodedProof: decodedProof,
+	}, nil
+}
+
+// DecodeProof attempts to decode a given proof file that claims to be
+// anchored at the specified genesis point.
+func (r *rpcServer) DecodeProof(ctx context.Context,
+	in *tarorpc.DecodeProofRequest) (*tarorpc.DecodeProofResponse, error) {
+
+	if len(in.RawProof) == 0 {
+		return nil, fmt.Errorf("proof file must be specified")
+	}
+
+	var proofFile proof.File
+	if err := proofFile.Decode(bytes.NewReader(in.RawProof)); err != nil {
+		return nil, fmt.Errorf("unable to decode proof file: %w", err)
+	}
+
+	var proofIdx int64
+	switch filter := in.Filter.(type) {
+	case *tarorpc.DecodeProofRequest_ProofIndex:
+		proofIdx = filter.ProofIndex
+	default:
+		proofIdx = -1
+	}
+
+	decodedProof, err := marshalProofFile(proofFile, proofIdx)
+	if err != nil {
+		return nil, fmt.Errorf("unable to marshal proof: %w", err)
+	}
+
+	return &tarorpc.DecodeProofResponse{
+		DecodedProof: decodedProof,
+	}, nil
+}
+
+// marshalProofFile turns a proof file into an RPC DecodedProof.
+func marshalProofFile(proofFile proof.File,
+	index int64) (*tarorpc.DecodedProof, error) {
+
+	var (
+		decodedProof *proof.Proof
+		err          error
+	)
+
+	// Default to latest proof
+	if index < 0 {
+		decodedProof, err = proofFile.LastProof()
+		if err != nil {
+			return nil, err
+		}
+		index = int64(proofFile.NumProofs() - 1)
+	} else {
+		decodedProof, err = proofFile.ProofAt(uint32(index))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var (
+		finalAsset     = decodedProof.Asset
+		rpcAsset       = &tarorpc.Asset{}
+		anchorOutpoint = wire.OutPoint{
+			decodedProof.AnchorTx.TxHash(),
+			decodedProof.InclusionProof.OutputIndex,
+		}
+		txMerkleProof  = decodedProof.TxMerkleProof
+		inclusionProof = decodedProof.InclusionProof
+	)
+
+	var txMerkleProofBuf bytes.Buffer
+	if err := txMerkleProof.Encode(&txMerkleProofBuf); err != nil {
+		return nil, fmt.Errorf("unable to encode serialized Bitcoin "+
+			"merkle proof: %w", err)
+	}
+	txMerkleProofBytes := txMerkleProofBuf.Bytes()
+
+	var inclusionProofBuf bytes.Buffer
+	if err := inclusionProof.Encode(&inclusionProofBuf); err != nil {
+		return nil, fmt.Errorf("unable to encode inclusion proof: %w",
+			err)
+	}
+	inclusionProofBytes := inclusionProofBuf.Bytes()
+
+	var exclusionProofs [][]byte
+	for _, exclusionProof := range decodedProof.ExclusionProofs {
+		var exclusionProofBuf bytes.Buffer
+		if err := exclusionProof.Encode(&exclusionProofBuf); err != nil {
+			return nil, fmt.Errorf("unable to encode exclusion proofs: %w",
```

Comment:
nit: need to break before word "proofs" to fit properly (with tab with set to 8 spaces).